src/components/generic/form.tsx

```diff
@@ -0,0 +1,334 @@
+
+  /**
+   * Method for setting date
+   *
+   * @param fieldName field name
+   */
+  const renderDatePicker = (fieldName: string) => {
+    const value = getFieldValue(fieldName);
+    return (
+      <LocalizationProvider dateAdapter={ AdapterDateFns } locale={ fiLocale }>
+        <DatePicker
+          value={ value ? new Date(value as string) : null }
+          onChange={ handleDateChange }
+          views={["day", "month", "year"]}
+          renderInput={ params =>
+            <TextField label={ strings.formComponent.dateTimePicker } { ...params }/>
+          }
+        />
+      </LocalizationProvider>
+    );
+  };
+
+  /**
+   * Method for setting datetime
+   */
+  const renderDatetimePicker = (fieldName: string) => {
+    const value = getFieldValue(fieldName);
+    return (
+      <LocalizationProvider dateAdapter={ AdapterDateFns } locale={ fiLocale }>
+        <DateTimePicker
+          value={ value ? new Date(value as string) : null }
+          onChange={ handleDateTimeChange }
+          renderInput={ params =>
+            <TextField label={ strings.formComponent.dateTimePicker } { ...params }/>
+          }
+        />
+      </LocalizationProvider>
+    );
+  };
+
+  /**
+   * Creates url with default format for accessing uploaded file
+   * TODO: Implement once keycloak is implemented
+   * @param id fileRef id
+   */
+  const createDefaultFileUrl = (id: string) => {
+    // return `${Api.createDefaultUploadUrl()}?fileRef=${id}`;
+  }
+
+  /**
+   * Performs file upload request
+   * 
+   * @param fieldName field name
+   * @param file file to upload
+   * @param path upload path
+   */
+  const doUpload = (fieldName: string, file: File, path: string) => {
+    setUploadingFields([...uploadingFields, fieldName]);
+    const data = new FormData();
+    data.append("file", file);
+    fetch(path, {
+      method: "POST",
+      body: data
+    })
+    .then(res => res.json())
+    .then((data) => {
+      let currentFiles = getFieldValue(fieldName);
+      if (!currentFiles) {
+        currentFiles = { files: [] };
+      }
+      const value = {
+        id: data.fileRef,
+        persisted: false,
+        name: data.fileName,
+        url: createDefaultFileUrl(data.fileRef)
+      } as FileFieldValueItem;
+      (currentFiles as FileFieldValue).files.push(value);
+      setFieldValue(fieldName, {...currentFiles as FileFieldValue});
+      setUploadingFields([...uploadingFields.filter(f => f !== fieldName)]);
+    })
+    .catch((e) => {
+      setUploadingFields([...uploadingFields.filter(f => f !== fieldName)]);
+    })
+  }
+
+  /**
+   * Method for uploading a file
+   *
+   * @param file file
+   * @param path path
+   */
+  const uploadFile = (fieldName: string, files: FileList | File, path: string) => {
+    if (files instanceof FileList) {
+      for (let i = 0; i < files.length; i++) {
+        let item = files.item(i);
+        if (item) {
+          doUpload(fieldName, item, path);
+        }
+      }
```

Comment:
FileList does not support forEach since it's not an array. It can be converted but I wouldn't do it before it can be properly tested in form.

Reply:
That might be true and it can be looked into in another issue once the logic can be tested (we have test form with upload files field). At the moment this code is just copy-pasted from metaform-ui project and tested working there.